src/tensor/index.rs

```diff
@@ -0,0 +1,304 @@
+        let c = index.2.into();
+        self.indexer(&[a, b, c])
+    }
+}
+
+impl<A, B, C, D> IndexOp<(A, B, C, D)> for Tensor
+where
+    A: Into<TensorIndexer>,
+    B: Into<TensorIndexer>,
+    C: Into<TensorIndexer>,
+    D: Into<TensorIndexer>,
+{
+    fn i(&self, index: (A, B, C, D)) -> Tensor {
+        let a = index.0.into();
+        let b = index.1.into();
+        let c = index.2.into();
+        let d = index.3.into();
+        self.indexer(&[a, b, c, d])
+    }
+}
+
+impl<A, B, C, D, E> IndexOp<(A, B, C, D, E)> for Tensor
+where
+    A: Into<TensorIndexer>,
+    B: Into<TensorIndexer>,
+    C: Into<TensorIndexer>,
+    D: Into<TensorIndexer>,
+    E: Into<TensorIndexer>,
+{
+    fn i(&self, index: (A, B, C, D, E)) -> Tensor {
+        let a = index.0.into();
+        let b = index.1.into();
+        let c = index.2.into();
+        let d = index.3.into();
+        let e = index.4.into();
+        self.indexer(&[a, b, c, d, e])
+    }
+}
+
+impl<A, B, C, D, E, F> IndexOp<(A, B, C, D, E, F)> for Tensor
+where
+    A: Into<TensorIndexer>,
+    B: Into<TensorIndexer>,
+    C: Into<TensorIndexer>,
+    D: Into<TensorIndexer>,
+    E: Into<TensorIndexer>,
+    F: Into<TensorIndexer>,
+{
+    fn i(&self, index: (A, B, C, D, E, F)) -> Tensor {
+        let a = index.0.into();
+        let b = index.1.into();
+        let c = index.2.into();
+        let d = index.3.into();
+        let e = index.4.into();
+        let f = index.5.into();
+        self.indexer(&[a, b, c, d, e, f])
+    }
+}
+
+impl<A, B, C, D, E, F, G> IndexOp<(A, B, C, D, E, F, G)> for Tensor
+where
+    A: Into<TensorIndexer>,
+    B: Into<TensorIndexer>,
+    C: Into<TensorIndexer>,
+    D: Into<TensorIndexer>,
+    E: Into<TensorIndexer>,
+    F: Into<TensorIndexer>,
+    G: Into<TensorIndexer>,
+{
+    fn i(&self, index: (A, B, C, D, E, F, G)) -> Tensor {
+        let a = index.0.into();
+        let b = index.1.into();
+        let c = index.2.into();
+        let d = index.3.into();
+        let e = index.4.into();
+        let f = index.5.into();
+        let g = index.6.into();
+        self.indexer(&[a, b, c, d, e, f, g])
+    }
+}
+
+impl Tensor {
+    fn indexer(&self, index_spec: &[TensorIndexer]) -> Tensor {
+        use std::ops::Bound::*;
+        use TensorIndexer::*;
+
+        // Make sure n. non-newaxis does not exceed n. of dimensions
+        let n_newaxis = index_spec.iter().fold(0, |mut count, spec| {
+            if spec == &InsertNewAxis {
+                count += 1
+            }
+            count
+        });
+
+        assert!(
+            index_spec.len() <= self.size().len() + n_newaxis,
+            format!(
+                "too many indices for tensor of dimension {}",
+                self.size().len()
+            )
+        );
+
+        // Apply indexing from left to right
+        let mut curr_tensor = self.shallow_clone();
+        let mut curr_idx: i64 = 0;
+
+        for (_spec_idx, spec) in index_spec.iter().enumerate() {
```

Comment:
If `_spec_idx` is not used, maybe don't use `enumerate()` ?

Reply:
It was left for (maybe) advanced indexing in the future. I remove it for now.